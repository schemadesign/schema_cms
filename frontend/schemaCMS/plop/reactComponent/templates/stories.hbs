import React from 'react';
import { storiesOf } from '@storybook/react';

import { {{ pascalCase name }} } from './{{ camelCase name }}.component';

export const defaultProps = {};

storiesOf('{{ pascalCase name }}', module).add('Default', () => <{{ pascalCase name }} {...defaultProps} />);
