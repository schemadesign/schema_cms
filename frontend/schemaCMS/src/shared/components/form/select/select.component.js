import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'schemaUI';

import { Container } from './select.styles';

const { Label, Select: SelectElement } = Form;

export class Select extends PureComponent {
  static propTypes = {
    defaultOption: PropTypes.object,
    label: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    options: PropTypes.array.isRequired,
    onSelect: PropTypes.func.isRequired,
  };

  render() {
    const { defaultOption, label, name, options, onSelect, value } = this.props;
    const updatedOptions = options.map(option => ({ ...option, selected: option.value === value }));

    return (
      <Container>
        <Label>{label}</Label>
        <input type="hidden" id={name} name={name} value={value} />
        <SelectElement defaultOption={defaultOption} options={updatedOptions} onSelect={onSelect} />
      </Container>
    );
  }
}
