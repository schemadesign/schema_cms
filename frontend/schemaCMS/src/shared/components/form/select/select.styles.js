import styled, { css } from 'styled-components';
import { identity } from 'ramda';

import { styleWhenTrue } from '../../../utils/rendering';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding-bottom: 30px;
  position: relative;
  font-size: 18px;
  cursor: pointer;

  label {
    margin-bottom: 2px;
  }

  ${({ customStyles }) => customStyles};
`;

const getCenterIconStyles = styleWhenTrue(
  identity,
  css`
    top: 50%;
    right: 5px;
    transform: translateY(-50%);
    font-size: 0;
  `
);

export const IconContainer = styled.div`
  position: absolute;
  right: 0;
  top: 5px;

  ${({ centerIcon }) => getCenterIconStyles(centerIcon)}
`;
