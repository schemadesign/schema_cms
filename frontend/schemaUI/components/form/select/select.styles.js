import { defaultTheme } from '../../../utils/theme';

const ITEM_HEIGHT = 30;
const PADDING = 7;

export const getStyles = (theme = defaultTheme) => ({
  containerStyles: {
    display: 'flex',
    width: '100%',
    color: theme.text,
  },
  getSelectStyle: (hidden = false) => ({
    display: hidden ? 'none' : 'flex',
  }),
  selectWrapperStyles: {
    display: 'flex',
    position: 'relative',
    width: '100%',
    cursor: 'pointer',
  },
  customOptionStyle: index => ({
    display: 'flex',
    cursor: 'pointer',
    alignItems: 'center',
    minHeight: '60px',
    padding: '0 10px',
    border: `1px solid  ${theme.select.border}`,
    borderTop: !index ? `1px solid  ${theme.select.border}` : 'none',
    color: theme.secondaryText,
  }),
  selectedOptionsStyles: {
    display: 'flex',
    height: `${ITEM_HEIGHT}px`,
    cursor: 'pointer',
    width: '100%',
  },
  selectedOptionStyles: {
    display: 'flex',
    cursor: 'pointer',
    height: `${ITEM_HEIGHT}px`,
    alignItems: 'center',
  },
  optionListStyles: menuOpen => ({
    display: menuOpen ? 'flex' : 'none',
    maxHeight: '175px',
    overflow: menuOpen ? 'auto' : 'hidden',
    flexDirection: 'column',
    position: 'absolute',
    width: '100%',
    top: `${ITEM_HEIGHT + PADDING}px`,
    left: '0',
    zIndex: 2,
    backgroundColor: theme.select.background,
  }),
  hoverStyles: {
    backgroundColor: theme.text,
    color: theme.background,
  },
});
