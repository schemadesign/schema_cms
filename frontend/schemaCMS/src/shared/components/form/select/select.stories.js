import React, { PureComponent } from 'react';
import { storiesOf } from '@storybook/react';

import { Select } from './select.component';
import { withTheme } from '../../../../.storybook/decorators';

export const defaultProps = {
  label: 'Selector',
  name: 'selectField',
  options: [{ value: 'one', label: 'One' }, { value: 'two', label: 'Two' }, { value: 'three', label: 'Three' }],
};

class TestComponent extends PureComponent {
  state = { selectedOption: 'one' };

  handleSelect = ({ value: selectedOption }) => this.setState({ selectedOption: selectedOption });

  render() {
    const { selectedOption } = this.state;

    return (
      <div>
        <Select {...defaultProps} value={selectedOption} onSelect={this.handleSelect} />
        <br />
        <br />
        Value: {selectedOption}
      </div>
    );
  }
}

storiesOf('Shared Components|Form/Select', module)
  .addDecorator(withTheme())
  .add('Default', () => <TestComponent />);
