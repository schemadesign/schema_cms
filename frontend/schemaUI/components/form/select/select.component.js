import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { getStyles } from './select.styles';
import { withStyles } from '../../styles/withStyles';

export class SelectComponent extends PureComponent {
  static propTypes = {
    native: PropTypes.bool,
    defaultOption: PropTypes.object,
    options: PropTypes.array.isRequired,
    onSelect: PropTypes.func.isRequired,
  };

  static defaultProps = {
    native: false,
  };

  state = {
    isMenuOpen: false,
  };

  handleOptionClick = ({ target: { value } }) => {
    const selectedOption = this.props.options.find(option => option.value === value);
    this.props.onSelect(selectedOption);

    this.setState({
      isMenuOpen: false,
    });
  };

  toggleMenu = () => this.setState({ isMenuOpen: !this.state.isMenuOpen });

  renderSelectedOption = ({ selectedOptionStyles }) => {
    const selectedOption = this.props.options.find(option => option.selected) || this.props.options[0];

    return <div style={selectedOptionStyles}>{selectedOption.label}</div>;
  };

  renderOptions = ({ value, label, selected }, index, { customOptionStyle }, id = '') => {
    const eventObj = {
      target: {
        value: value,
      },
    };

    return (
      <div
        key={index}
        id={`select-item-${index}`}
        style={customOptionStyle(index)}
        onClick={() => this.handleOptionClick(eventObj)}
      >
        {label}
      </div>
    );
  };

  renderNativeOptions = ({ value, label, selected }, index) => (
    <option key={index} value={value}>
      {label}
    </option>
  );

  renderNativeSelect = ({ hidden = false, restStyles: { getSelectStyle } }, restProps) => {
    const { options } = this.props;
    const onChange = !!hidden ? Function.prototype : this.handleOptionClick;

    return (
      <select style={getSelectStyle(hidden)} onChange={onChange} {...restProps}>
        {options.map(this.renderNativeOptions)}
      </select>
    );
  };

  renderCustomSelect = ({ selectWrapperStyles, selectedOptionsStyles, optionListStyles, ...restStyles }, props) => {
    const { id } = props;

    return (
      <div id={id} style={selectWrapperStyles}>
        <div style={selectedOptionsStyles} onClick={this.toggleMenu}>
          {this.renderSelectedOption(restStyles)}
        </div>
        <div style={optionListStyles(this.state.isMenuOpen)}>
          {this.props.options.map((item, index) => this.renderOptions(item, index, restStyles, id))}
        </div>
      </div>
    );
  };

  render() {
    const { native, theme, ...restProps } = this.props;
    const { containerStyles, ...restStyles } = getStyles(theme);

    return (
      <div style={containerStyles}>
        {native ? this.renderNativeSelect({ restStyles }, restProps) : this.renderCustomSelect(restStyles, restProps)}
      </div>
    );
  }
}

export const Select = withStyles(SelectComponent);
