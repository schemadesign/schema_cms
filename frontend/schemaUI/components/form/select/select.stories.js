import React, { PureComponent } from 'react';
import { storiesOf } from '@storybook/react';
import PropTypes from 'prop-types';

import { SelectComponent } from './select.component';
import { withTheme } from '../../../.storybook/decorators';

const options = [
  { value: 'dolphin', label: 'Dolphin' },
  { value: 'dog', label: 'Dog' },
  { value: 'parrot', label: 'Parrot' },
  { value: 'cat', label: 'Cat' },
  { value: 'alligator', label: 'Alligator' },
  { value: 'spider', label: 'Spider' },
];

class TestComponent extends PureComponent {
  static propTypes = {
    native: PropTypes.bool,
  };

  state = { selectedOption: {}, native: false };

  onSelect = selectedOption => this.setState({ selectedOption: { ...selectedOption, selected: true } });

  render() {
    const selectedOption = this.state.selectedOption;

    const native = this.props.native || this.state.native;

    const parsedOptions = options.map(option => (option.label === selectedOption.label ? selectedOption : option));

    return (
      <div>
        <SelectComponent options={parsedOptions} onSelect={this.onSelect} native={native} />
        <br />
        <br />
        Selected: {selectedOption.label}
        <br />
        Value: {selectedOption.value}
      </div>
    );
  }
}

class TestComponentWithPlaceholder extends PureComponent {
  state = { selectedOption: '' };

  onSelect = selectedOption => this.setState({ selectedOption: { ...selectedOption, selected: true } });

  render() {
    const selectedOption = this.state.selectedOption;
    const parsedOptions = options.map(option => (option.label === selectedOption.label ? selectedOption : option));

    return (
      <div>
        <SelectComponent options={parsedOptions} onSelect={this.onSelect} placeholder="Select one" />
        <br />
        <br />
        Selected: {selectedOption.label}
        <br />
        Value: {selectedOption.value}
      </div>
    );
  }
}

storiesOf('Form/Select', module)
  .addDecorator(withTheme())
  .add('Default', () => <TestComponent />)
  .add('With Placeholder', () => <TestComponentWithPlaceholder />)
  .add('Native Select', () => <TestComponent native />);
