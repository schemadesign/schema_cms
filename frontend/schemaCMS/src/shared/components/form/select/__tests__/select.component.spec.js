import React from 'react';
import { shallow } from 'enzyme';

import { Select } from '../select.component';
import { defaultProps } from '../select.stories';

describe('Select: Component', () => {
  const component = props => <Select {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render({
      value: 'one',
      onSelect: Function.prototype,
    });
    global.expect(wrapper).toMatchSnapshot();
  });
});
