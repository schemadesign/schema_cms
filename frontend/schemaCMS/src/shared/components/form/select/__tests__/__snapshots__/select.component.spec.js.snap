// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select: Component should render correctly 1`] = `
<styled.div>
  <Label
    customStyles={Object {}}
  >
    Selector
  </Label>
  <input
    id="selectField"
    name="selectField"
    type="hidden"
  />
  <Select
    native={false}
    onSelect={[Function]}
    options={
      Array [
        Object {
          "label": "One",
          "selected": false,
          "value": "one",
        },
        Object {
          "label": "Two",
          "selected": false,
          "value": "two",
        },
        Object {
          "label": "Three",
          "selected": false,
          "value": "three",
        },
      ]
    }
  />
</styled.div>
`;
